src/codegen/spirv.zig

```diff
@@ -84,23 +385,197 @@ pub const SPIRVModule = struct {
+                const param_type_id = self.types.get(tv.ty.fnParamType(i)).?;
+                const arg_result_id = self.spv.allocResultId();
+                try writeInstruction(&self.spv.fn_decls, .OpFunctionParameter, &[_]u32{ param_type_id, arg_result_id });
+                self.args.appendAssumeCapacity(arg_result_id);
+            }
+
+            // TODO: This could probably be done in a better way...
+            const root_block_id = self.spv.allocResultId();
+            _ = try writeInstruction(&self.spv.fn_decls, .OpLabel, &[_]u32{root_block_id});
+            try self.genBody(func_payload.data.body);
+
+            try writeInstruction(&self.spv.fn_decls, .OpFunctionEnd, &[_]u32{});
+        } else {
+            return self.fail(.{.node_offset = 0}, "TODO: SPIR-V backend: generate decl type {}", .{ tv.ty.zigTypeTag() });
+        }
+    }
+
+    fn genBody(self: *DeclGen, body: ir.Body) !void {
+        for (body.instructions) |inst| {
+            const maybe_result_id = try self.genInst(inst);
+            if (maybe_result_id) |result_id|
+                try self.values.putNoClobber(inst, result_id);
         }
     }
+
+    fn genInst(self: *DeclGen, inst: *Inst) !?u32 {
+        return switch (inst.tag) {
+            .add, .addwrap => try self.genBinOp(inst.castTag(.add).?),
+            .sub, .subwrap => try self.genBinOp(inst.castTag(.sub).?),
+            .mul, .mulwrap => try self.genBinOp(inst.castTag(.mul).?),
+            .div => try self.genBinOp(inst.castTag(.div).?),
+            .bit_and => try self.genBinOp(inst.castTag(.bit_and).?),
+            .bit_or => try self.genBinOp(inst.castTag(.bit_or).?),
+            .xor => try self.genBinOp(inst.castTag(.xor).?),
+            .cmp_eq => try self.genBinOp(inst.castTag(.cmp_eq).?),
+            .cmp_neq => try self.genBinOp(inst.castTag(.cmp_neq).?),
+            .cmp_gt => try self.genBinOp(inst.castTag(.cmp_gt).?),
+            .cmp_gte => try self.genBinOp(inst.castTag(.cmp_gte).?),
+            .cmp_lt => try self.genBinOp(inst.castTag(.cmp_lt).?),
+            .cmp_lte => try self.genBinOp(inst.castTag(.cmp_lte).?),
+            .bool_and => try self.genBinOp(inst.castTag(.bool_and).?),
+            .bool_or => try self.genBinOp(inst.castTag(.bool_or).?),
+            .not => try self.genUnOp(inst.castTag(.not).?),
+            .arg => self.genArg(),
+            // TODO: Breakpoints won't be supported in SPIR-V, but the compiler seems to insert them
+            // throughout the IR.
+            .breakpoint => null,
+            .dbg_stmt => null,
+            .ret => self.genRet(inst.castTag(.ret).?),
+            .retvoid => self.genRetVoid(),
+            .unreach => self.genUnreach(),
+            else => self.fail(.{.node_offset = 0}, "TODO: SPIR-V backend: implement inst {}", .{inst.tag}),
+        };
+    }
+
+    fn genBinOp(self: *DeclGen, inst: *Inst.BinOp) !u32 {
+        // TODO: Will lhs and rhs have the same type?
```

Comment:
Good question which I am happy to leave as a TODO comment until AIR tags are properly documented with doc comments. (Answer is yes except possibly for shifts which I want to double check)

Reply:
Ah, shifts, i forgot about those. I'll leave it like this for now until shifts are there.